docs/sdk/core/examples/app2_test.go

```diff
@@ -0,0 +1,83 @@
+package app
+
+import (
+	"testing"
+	"github.com/tendermint/tendermint/crypto"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/stretchr/testify/require"
+)
+
+// Test encoding of app2Tx is correct with both msg types
+func TestEncoding(t *testing.T) {
+	// Create privkeys and addresses
+	priv1 := crypto.GenPrivKeyEd25519()
+	priv2 := crypto.GenPrivKeyEd25519()
+	addr1 := priv1.PubKey().Address().Bytes()
+	addr2 := priv2.PubKey().Address().Bytes()
+
+	sendMsg := MsgSend{
+		From: addr1,
+		To: addr2,
+		Amount: sdk.Coins{{"testCoins", sdk.NewInt(100)}},
+	}
+
+	// Construct transaction
+	signBytes := sendMsg.GetSignBytes()
+	sig, err := priv1.Sign(signBytes)
+	if err != nil {
+		panic(err)
+	}
+
+	sendTxBefore := app2Tx{
+		Msg: sendMsg,
+		PubKey: priv1.PubKey(),
+		Signature: sig,
+	}
+
+	cdc := NewCodec()
+	testTxDecoder := tx2Decoder(cdc)
+
+	encodedSendTx, err := cdc.MarshalBinary(sendTxBefore)
+
+	require.Nil(t, err, "Error encoding sendTx")
+
+	var tx1 sdk.Tx
+	tx1, err = testTxDecoder(encodedSendTx)
+	require.Nil(t, err, "Error decoding sendTx")
+
+	sendTxAfter := tx1.(app2Tx)
+	
+	require.Equal(t, sendTxBefore, sendTxAfter, "Transaction changed after encoding/decoding")
+
+	issueMsg := MsgIssue{
+		Issuer: addr1,
+		Receiver: addr2,
+		Coin: sdk.Coin{"testCoin", sdk.NewInt(100)},
+	}
+
+	signBytes = issueMsg.GetSignBytes()
```

Comment:
can we deduplicate all this so we can easily check the encodings with multiple message types ?

Reply:
What do you mean by this?